## Pandas Basics

Pandas is a software library written for the Python programming language for data manipulation and analysis. In particular, it offers data structures and operations for manipulating numerical tables and time series. 

In [1]:
import pandas as pd

### 1.Reading in DataFrames From Files

The pandas.read_csv command reads a comma-separated values (csv) file into DataFrame.

_Using the read_csv command to load elections.csv file and Assign it as elections_

In [2]:
elections = pd.read_csv("elections.csv")
elections

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2    56.203927  
3    43.796073  
4    54.574789  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[182 rows x 6 columns]

### 2.Index with loc,iloc,and []

The Pandas library has a lot of “syntactic sugar”: Methods that are useful and lead to concise code, but not absolutely necessary for the library to function.  
- Examples: .head and .tail

In [3]:
elections.tail(5)

Year       Candidate        Party  Popular vote Result          %
177  2016      Jill Stein        Green       1457226   loss   1.073699
178  2020    Joseph Biden   Democratic      81268924    win  51.311515
179  2020    Donald Trump   Republican      74216154   loss  46.858542
180  2020    Jo Jorgensen  Libertarian       1865724   loss   1.177979
181  2020  Howard Hawkins        Green        405035   loss   0.255731

### (1) loc : select a group of rows and columns by label(s).

Labels are fundamentally:
- The labels are the bolded text to the top and left of our dataframe.  
- Row labels shown: 177, 178, 179, 180, 181  
- Column labels: Year, Candidate, Party, Popular vote, Result, %

#### Arguments to loc can be:  
- A list
- A slice(syntax is inclusive of the right hand side of the slice)
- A value

_Argument as a list: The following command example selects the values of Year, Candidate, Result columns in 87th,125th,and 179th rows._

In [4]:
elections.loc[[87,125,179], ["Year","Candidate","Result"]]

Year       Candidate Result
87   1932  Herbert Hoover   loss
125  1976  Roger MacBride   loss
179  2020    Donald Trump   loss

_Argument as a slice: The following command example selects the values of Party column to Result column in 87th,125th,and 179th rows._

In [5]:
elections.loc[[87,125,179],'Party':'Result']

Party  Popular vote Result
87    Republican      15761254   loss
125  Libertarian        172557   loss
179   Republican      74216154   loss

_Argument as a value: The following command example selects the values of Candidate column in 87th,125th,and 179th rows._

In [6]:
elections.loc[[87,125,179],['Candidate']]

Candidate
87   Herbert Hoover
125  Roger MacBride
179    Donald Trump

_If you want all rows, but only some columns, you can use : for the left argument._

In [7]:
elections.loc[:,['Year','Popular vote']]

Year  Popular vote
0    1824        151271
1    1824        113142
2    1828        642806
3    1828        500897
4    1832        702735
..    ...           ...
177  2016       1457226
178  2020      81268924
179  2020      74216154
180  2020       1865724
181  2020        405035

[182 rows x 2 columns]

### (2) iloc : select a group of rows and columns numbers by index.

#### Arguments to loc can be:  
- A list
- A slice(syntax is exclusive of the right hand side of the slice)
- A value

_Argument as a list: The following command example selects the values of first and second columns in 17th,25th,and 79th rows._

In [8]:
elections.iloc[[17,25,79], [0, 1]]

Year             Candidate
17  1852       Franklin Pierce
25  1860  John C. Breckinridge
79  1920        Warren Harding

_Argument as a slice: The following command example selects the values of first and second columns(exclusively the third column) in 17th,25th,and 79th rows._

In [9]:
elections.iloc[[17, 25, 79], 0:2]

Year             Candidate
17  1852       Franklin Pierce
25  1860  John C. Breckinridge
79  1920        Warren Harding

_Argument as a vlue: The following command example selects the values of first column in 17th,25th,and 79th rows._

In [10]:
elections.iloc[[17,25,79],[0]]

Year
17  1852
25  1860
79  1920

### (3) []: selection operator

#### [] only takes one argument, which may be:  
- A slice of row numbers.
- A list of columns.
- A single column label.

In [11]:
# A slice of row numbers (exclusively the right side)
elections[0:3]

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927

In [12]:
# A list of columns
elections[['Party','Result']].head()

Party Result
0  Democratic-Republican   loss
1  Democratic-Republican    win
2             Democratic    win
3    National Republican   loss
4             Democratic    win

In [13]:
# A column label
elections["Candidate"].head(3)

0       Andrew Jackson
1    John Quincy Adams
2       Andrew Jackson
Name: Candidate, dtype: object

The above formatting is a bit ugly because the output has only one column. To fix this formatting, we will use the dataframe format instead of series and introduce more about fundamental data structures in pandas library later.

In [14]:
#Select even a single column by a list
elections[["Candidate"]].head(3)

Candidate
0     Andrew Jackson
1  John Quincy Adams
2     Andrew Jackson

In summary, [] is much more common in real world compared to loc and iloc especially when selecting columns.

### 3.DataFrames,Series, and Indicies

In [15]:
#output the type of elections
type(elections)

pandas.core.frame.DataFrame

In [16]:
#output the type of elections
type(elections['Year'])

pandas.core.series.Series

#### There are three fundamental data structures in pandas:  
- Data Frame: 2D data tabular data.
- Series: 1D data, which is usually regarded as a column of data.
- Index: A sequence of row labels.

We can think of a Data Frame as a collection of Series that all share the same Index.
e.g.:Candidate, Party, %, Year, and Result Series all share an Index from  0 to 5.

An Index can also be non-numetric and have a name, e.g."Candidate"

In [17]:
elections_by_candidate = elections.set_index('Candidate')
elections_by_candidate.head()

Year                  Party  Popular vote Result          %
Candidate                                                                     
Andrew Jackson     1824  Democratic-Republican        151271   loss  57.210122
John Quincy Adams  1824  Democratic-Republican        113142    win  42.789878
Andrew Jackson     1828             Democratic        642806    win  56.203927
John Quincy Adams  1828    National Republican        500897   loss  43.796073
Andrew Jackson     1832             Democratic        702735    win  54.574789

The row labels that constitute an index do not have to be unique.

In [18]:
elections_by_candidate.index

Index(['Andrew Jackson', 'John Quincy Adams', 'Andrew Jackson',
       'John Quincy Adams', 'Andrew Jackson', 'Henry Clay', 'William Wirt',
       'Hugh Lawson White', 'Martin Van Buren', 'William Henry Harrison',
       ...
       'Darrell Castle', 'Donald Trump', 'Evan McMullin', 'Gary Johnson',
       'Hillary Clinton', 'Jill Stein', 'Joseph Biden', 'Donald Trump',
       'Jo Jorgensen', 'Howard Hawkins'],
      dtype='object', name='Candidate', length=182)

In [19]:
elections_by_candidate.loc[['Donald Trump'],]

Year       Party  Popular vote Result          %
Candidate                                                     
Donald Trump  2016  Republican      62984828    win  46.407862
Donald Trump  2020  Republican      74216154   loss  46.858542

#### Getting a DataFrame Rather Than a Series
Suppose we want our single column back as a DataFrame. Two approaches:  
- Use Series.to_frame().  
- Provide a list containing the single column of interest. It is like double braces but this approach actually provides a list to the [] operator.

In [20]:
#Approach 1
elections['Candidate'].tail().to_frame()

Candidate
177      Jill Stein
178    Joseph Biden
179    Donald Trump
180    Jo Jorgensen
181  Howard Hawkins

In [21]:
#Approach 2
elections[['Candidate']].tail()

Candidate
177      Jill Stein
178    Joseph Biden
179    Donald Trump
180    Jo Jorgensen
181  Howard Hawkins

#### Retrieve Row and Column Labels

In [22]:
#For row labels, use DataFrame.index
elections.index

RangeIndex(start=0, stop=182, step=1)

In [23]:
#For column labels. use DataFrame.columns
elections.columns

Index(['Year', 'Candidate', 'Party', 'Popular vote', 'Result', '%'], dtype='object')

### 4. Conditioanl Selection

Using boolean arrays is to select subset dataframes which satisfies the conditions.

In [24]:
#Output whether each row of Party is Democratic or not
elections["Party"] == 'Democratic'

0      False
1      False
2       True
3      False
4       True
       ...  
177    False
178     True
179    False
180    False
181    False
Name: Party, Length: 182, dtype: bool

In [25]:
#Select first five rows whose party is Democratic (when boolean array is true)
elections[elections["Party"] == 'Democratic'].head()

Year         Candidate       Party  Popular vote Result          %
2   1828    Andrew Jackson  Democratic        642806    win  56.203927
4   1832    Andrew Jackson  Democratic        702735    win  54.574789
8   1836  Martin Van Buren  Democratic        763291    win  52.272472
10  1840  Martin Van Buren  Democratic       1128854   loss  46.948787
13  1844        James Polk  Democratic       1339570    win  50.749477

Boolean Series can be combined using various operators, allowing filtering of results by multiple criteria.

In [26]:
elections[(elections["Result"] == 'win') & (elections['%']<45)]

Year          Candidate                  Party  Popular vote Result  \
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
23   1860    Abraham Lincoln             Republican       1855993    win   
70   1912     Woodrow Wilson             Democratic       6296284    win   
117  1968      Richard Nixon             Republican      31783783    win   
140  1992       Bill Clinton             Democratic      44909806    win   

             %  
1    42.789878  
23   39.699408  
70   41.933422  
117  43.565246  
140  43.118485

#### Alternatives to Boolean Array Selection

Pandas also provides many alternatives, for example:  
- .isin
- .str.startwith
- .query

In [27]:
#DataFrame.isin(values) check whether each element in the DataFrame is contained in the values
a_parties = ["Anti-Masonic", "American", "Anti-Monopoly", "American Independent"]
elections[elections["Party"].isin(a_parties)]

Year           Candidate                 Party  Popular vote Result  \
6    1832        William Wirt          Anti-Masonic        100715   loss   
22   1856    Millard Fillmore              American        873053   loss   
38   1884     Benjamin Butler         Anti-Monopoly        134294   loss   
115  1968      George Wallace  American Independent       9901118   loss   
119  1972     John G. Schmitz  American Independent       1100868   loss   
124  1976       Lester Maddox  American Independent        170274   loss   
126  1976  Thomas J. Anderson              American        158271   loss   

             %  
6     7.821583  
22   21.554001  
38    1.335838  
115  13.571218  
119   1.421524  
124   0.209640  
126   0.194862

In [28]:
#Series.str.startswith(pat) test if the start of each string element matches a pattern.
elections[elections["Party"].str.startswith("S")]

Year             Candidate                Party  Popular vote Result  \
25   1860  John C. Breckinridge  Southern Democratic        848019   loss   
58   1904        Eugene V. Debs            Socialist        402810   loss   
62   1908        Eugene V. Debs            Socialist        420852   loss   
66   1912        Eugene V. Debs            Socialist        901551   loss   
71   1916       Allan L. Benson            Socialist        590524   loss   
76   1920        Eugene V. Debs            Socialist        913693   loss   
85   1928         Norman Thomas            Socialist        267478   loss   
88   1932         Norman Thomas            Socialist        884885   loss   
92   1936         Norman Thomas            Socialist        187910   loss   
95   1940         Norman Thomas            Socialist        116599   loss   
102  1948         Norman Thomas            Socialist        139569   loss   
110  1956    T. Coleman Andrews       States' Rights        107929   loss   

             %  
25   18.138998  
58    2.985897  
62    2.850866  
66    6.004354  
71    3.194193  
76    3.428282  
85    0.728623  
88    2.236211  
92    0.412876  
95    0.234237  
102   0.286312  
110   0.174883

In [29]:
# DataFrame.query(expr) query the columns of a DataFrame with a boolean expression.
elections.query('Year >= 2000 and Party == "Democratic"')

Year        Candidate       Party  Popular vote Result          %
151  2000          Al Gore  Democratic      50999897   loss  48.491813
158  2004       John Kerry  Democratic      59028444   loss  48.306775
162  2008     Barack Obama  Democratic      69498516    win  53.023510
168  2012     Barack Obama  Democratic      65915795    win  51.258484
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515

#### Handy Utility Functions

In [30]:
#size: return an int representing the number of elements in this object
elections.size

1092

In [31]:
#shape: return a tuple representing the dimensionality of the DataFrame
elections.shape

(182, 6)

In [32]:
#The describe function generates descriptive statistics of the DataFrame.
#Descriptive statistics include those that summarize the central tendency, dispersion and shape of a dataset’s distribution.
elections.describe()

Year  Popular vote           %
count   182.000000  1.820000e+02  182.000000
mean   1934.087912  1.235364e+07   27.470350
std      57.048908  1.907715e+07   22.968034
min    1824.000000  1.007150e+05    0.098088
25%    1889.000000  3.876395e+05    1.219996
50%    1936.000000  1.709375e+06   37.677893
75%    1988.000000  1.897775e+07   48.354977
max    2020.000000  8.126892e+07   61.344703

In [33]:
#sample(n): return a random sample of items from an axis of object.
# n : number of sampled items 
elections.sample(5)

Year       Candidate                 Party  Popular vote Result  \
138  1988        Ron Paul           Libertarian        431750   loss   
132  1984  David Bergland           Libertarian        228111   loss   
134  1984  Walter Mondale            Democratic      37577352   loss   
36   1880  James Garfield            Republican       4453337    win   
115  1968  George Wallace  American Independent       9901118   loss   

             %  
138   0.472660  
132   0.247245  
134  40.729429  
36   48.369234  
115  13.571218

In [36]:
# value_counts: return a Series containing counts of unique values
# The resulting object will be in descending order so that the first element is the most frequently-occurring element. 
# Excludes NA values by default.
elections[['Party']].value_counts().head()

Party      
Democratic     47
Republican     41
Libertarian    12
Prohibition    11
Socialist      10
dtype: int64

In [35]:
# sort_values: sort by the values along the specified 
elections[elections['Result'] == 'win'].sort_values(by = 'Popular vote', ascending = False).head()

Year       Candidate       Party  Popular vote Result          %
178  2020    Joseph Biden  Democratic      81268924    win  51.311515
162  2008    Barack Obama  Democratic      69498516    win  53.023510
168  2012    Barack Obama  Democratic      65915795    win  51.258484
173  2016    Donald Trump  Republican      62984828    win  46.407862
157  2004  George W. Bush  Republican      62040610    win  50.771824